In [1]:
import pandas as pd

In [2]:
train_data = pd.read_csv('data/train.csv', index_col='datetime')
test_data = pd.read_csv('data/test.csv', index_col='datetime')

In [3]:
train_data.index = pd.to_datetime(train_data.index)
test_data.index = pd.to_datetime(test_data.index)

In [4]:
train_data

season  holiday  workingday  weather   temp   atemp  \
datetime                                                                   
2011-01-01 00:00:00       1        0           0        1   9.84  14.395   
2011-01-01 01:00:00       1        0           0        1   9.02  13.635   
2011-01-01 02:00:00       1        0           0        1   9.02  13.635   
2011-01-01 03:00:00       1        0           0        1   9.84  14.395   
2011-01-01 04:00:00       1        0           0        1   9.84  14.395   
...                     ...      ...         ...      ...    ...     ...   
2012-12-19 19:00:00       4        0           1        1  15.58  19.695   
2012-12-19 20:00:00       4        0           1        1  14.76  17.425   
2012-12-19 21:00:00       4        0           1        1  13.94  15.910   
2012-12-19 22:00:00       4        0           1        1  13.94  17.425   
2012-12-19 23:00:00       4        0           1        1  13.12  16.665   

                     humidity  windspeed  casual  registered  count  
datetime                                                             
2011-01-01 00:00:00        81     0.0000       3          13     16  
2011-01-01 01:00:00        80     0.0000       8          32     40  
2011-01-01 02:00:00        80     0.0000       5          27     32  
2011-01-01 03:00:00        75     0.0000       3          10     13  
2011-01-01 04:00:00        75     0.0000       0           1      1  
...                       ...        ...     ...         ...    ...  
2012-12-19 19:00:00        50    26.0027       7         329    336  
2012-12-19 20:00:00        57    15.0013      10         231    241  
2012-12-19 21:00:00        61    15.0013       4         164    168  
2012-12-19 22:00:00        61     6.0032      12         117    129  
2012-12-19 23:00:00        66     8.9981       4          84     88  

[10886 rows x 11 columns]

In [24]:
test_data

season  holiday  workingday  weather   temp   atemp  \
datetime                                                                   
2011-01-20 00:00:00       1        0           1        1  10.66  11.365   
2011-01-20 01:00:00       1        0           1        1  10.66  13.635   
2011-01-20 02:00:00       1        0           1        1  10.66  13.635   
2011-01-20 03:00:00       1        0           1        1  10.66  12.880   
2011-01-20 04:00:00       1        0           1        1  10.66  12.880   
...                     ...      ...         ...      ...    ...     ...   
2012-12-31 19:00:00       1        0           1        2  10.66  12.880   
2012-12-31 20:00:00       1        0           1        2  10.66  12.880   
2012-12-31 21:00:00       1        0           1        1  10.66  12.880   
2012-12-31 22:00:00       1        0           1        1  10.66  13.635   
2012-12-31 23:00:00       1        0           1        1  10.66  13.635   

                     humidity  windspeed  
datetime                                  
2011-01-20 00:00:00        56    26.0027  
2011-01-20 01:00:00        56     0.0000  
2011-01-20 02:00:00        56     0.0000  
2011-01-20 03:00:00        56    11.0014  
2011-01-20 04:00:00        56    11.0014  
...                       ...        ...  
2012-12-31 19:00:00        60    11.0014  
2012-12-31 20:00:00        60    11.0014  
2012-12-31 21:00:00        60    11.0014  
2012-12-31 22:00:00        56     8.9981  
2012-12-31 23:00:00        65     8.9981  

[6493 rows x 8 columns]

In [25]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 10886 entries, 2011-01-01 00:00:00 to 2012-12-19 23:00:00
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   season      10886 non-null  int64  
 1   holiday     10886 non-null  int64  
 2   workingday  10886 non-null  int64  
 3   weather     10886 non-null  int64  
 4   temp        10886 non-null  float64
 5   atemp       10886 non-null  float64
 6   humidity    10886 non-null  int64  
 7   windspeed   10886 non-null  float64
 8   casual      10886 non-null  int64  
 9   registered  10886 non-null  int64  
 10  count       10886 non-null  int64  
dtypes: float64(3), int64(8)
memory usage: 1020.6 KB


In [26]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 6493 entries, 2011-01-20 00:00:00 to 2012-12-31 23:00:00
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   season      6493 non-null   int64  
 1   holiday     6493 non-null   int64  
 2   workingday  6493 non-null   int64  
 3   weather     6493 non-null   int64  
 4   temp        6493 non-null   float64
 5   atemp       6493 non-null   float64
 6   humidity    6493 non-null   int64  
 7   windspeed   6493 non-null   float64
dtypes: float64(3), int64(5)
memory usage: 456.5 KB


확인 작업
1) holiday와 workingday 둘 중 하나만 1이고 다른 하나는 0인지 확인하기.
2) casual + registered == count인지 확인하기

In [27]:
# 1) holiday & workingday
condition = (train_data['holiday'] + train_data['workingday']) == 1
all_valid = condition.all()
print(all_valid)

invalid_rows = train_data[(train_data['holiday'] + train_data['workingday']) != 1]
invalid_count = invalid_rows.shape[0]
print(invalid_count)

False
3163


1-1. 그럼 invalid_rows는 holiday와 workingday 모두 0인 것인지 확인하기.

In [28]:
condition1 = (invalid_rows['holiday'] + invalid_rows['workingday']) == 0
all_valid1 = condition1.all()
print(all_valid1)

True


그렇다면 holiday도 0이고, workingday도 0인 3163개의 행들은? 

공휴일도 아니고 workingday도 아닌 날은 토요일로 볼 수도 있겠다. &rarr; 실제로 그런지 확인

In [29]:
invalid_rows.index

DatetimeIndex(['2011-01-01 00:00:00', '2011-01-01 01:00:00',
               '2011-01-01 02:00:00', '2011-01-01 03:00:00',
               '2011-01-01 04:00:00', '2011-01-01 05:00:00',
               '2011-01-01 06:00:00', '2011-01-01 07:00:00',
               '2011-01-01 08:00:00', '2011-01-01 09:00:00',
               ...
               '2012-12-16 14:00:00', '2012-12-16 15:00:00',
               '2012-12-16 16:00:00', '2012-12-16 17:00:00',
               '2012-12-16 18:00:00', '2012-12-16 19:00:00',
               '2012-12-16 20:00:00', '2012-12-16 21:00:00',
               '2012-12-16 22:00:00', '2012-12-16 23:00:00'],
              dtype='datetime64[ns]', name='datetime', length=3163, freq=None)

In [31]:
notholiday_notworkingday = invalid_rows.index.day_name()
notholiday_notworkingday.value_counts()

datetime
Saturday    1584
Sunday      1579
Name: count, dtype: int64

확인 결과 holiday도 0이고 workingday도 0인 행들은 토요일 아니면 일요일이었다.

In [35]:
weekend_rows = train_data[(train_data.index.dayofweek==5) | (train_data.index.dayofweek==6)]
weekend_rows.shape

(3163, 11)

결론 ) 확인결과 모든 토요일과 일요일은 workingday = 0, holiday = 0 이었다.
그대로 두고 진행

In [36]:
#count확인
condition2 = train_data['count'] == (train_data['casual'] + train_data['registered'])
print(condition2.all())

True


In [15]:
train_x = train_data.iloc[:,:-3]
train_y = train_data.iloc[:,-1]